stone/target/js_types.py

```diff
@@ -159,7 +161,35 @@ def _generate_struct(self, struct_type, extra_parameters = [], nameOverride = No
         Emits a JSDoc @typedef for a struct.
         """
         self._emit_jsdoc_header(struct_type.doc)
-        self.emit(' * @typedef {Object} %s' % (nameOverride if nameOverride else fmt_type(struct_type)))
+        self.emit(' * @typedef {Object} %s' % (nameOverride if nameOverride else fmt_type_name(struct_type)))
+
+        # Some structs can explicitly list their subtypes. These structs have a .tag field that indicate
+        # which subtype they are.
+        if struct_type.is_member_of_enumerated_subtypes_tree():
+            if struct_type.has_enumerated_subtypes():
+                # This struct is the parent to multiple subtypes. Determine all of the possible
+                # values of the .tag property.
+                tag_values = []
+                for tags, _ in struct_type.get_all_subtypes_with_tags():
+                    for tag in tags:
+                        tag_values.append('"%s"' % tag)
+
+                jsdoc_tag_union = fmt_jsdoc_union(tag_values)
+                self.emit_wrapped_text('@property {%s} .tag - Tag identifying the subtype variant.' % jsdoc_tag_union)
+            else:
```

Comment:
This case is unnecessary. If one of the subtypes is directly referenced, then there will be no `.tag` attribute since it isn't ambiguous. An example of this is the `filesUpload` route. It will return a `FilesFileMetadata` object (you can only upload a file, not a folder), and there's no `.tag` attribute.

Also, we don't support multiple layers of inheritance with enumerated subtypes. It's always only a single root struct and one layer of children (we might expand the design later).